### ISSUE TITLE
False positive from dead_code lint when a repr(packed) struct is transmuted

### ISSUE THREAD

**Issue by sconybeare** [rust-lang/rust#79439]
```rust
#[repr(packed)]
struct Foo {
    a: u64,
}

#[repr(packed)]
struct Bar {
    x: u64, // this field is read as Foo.a after a transmute
}

fn main() {
    let bar = Bar { x: 5 };
    let foo: Foo = unsafe { std::mem::transmute(bar) };
    let value_of_bar_x = foo.a;
    println!("{}", value_of_bar_x);
}
```

([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=3eb6e9cfd464d8dd6cd5c3aa24c3cd35))

Output:

```
5

```

Errors:

```
   Compiling playground v0.0.1 (/playground)
warning: field is never read: `x`
 --> src/main.rs:8:5
  |
8 |     x: u64, // this field is read as Foo.a after a transmute
  |     ^^^^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: 1 warning emitted

    Finished dev [unoptimized + debuginfo] target(s) in 0.51s
     Running `target/debug/playground`

```
